### PR TITLE
feat: nullifier_registry contract functions

### PR DESCRIPTION
**Pull request by mubarak23**
PR for issue: #31

### CODE REVIEW

**Comment by mubarak23** [comment]
@manlikeHB you can pick up from here

**Comment by mubarak23** [comment]
@lana-shanghai this is ready for review

**Comment by mubarak23** [comment]
we also need to update our CI command for running the test to use 

`snforge test`

**Comment by mubarak23** [comment]
@lana-shanghai This PR is Ready for Review

File: contracts/src/contracts/nullifier_registry/nullifier_registry.cairo [hunk 1]
```diff
@@ -1 +1,294 @@
...
+        storage::{Map, StorageMapReadAccess, StorageMapWriteAccess}
+    };
+    use zkramp::contracts::nullifier_registry::interface::{
+        INullifierRegistry, INullifierRegistryDispatcher, INullifierRegistryDispatcherTrait
+    };
+
+    #[storage]
+    pub struct Storage {
+        // / map(index -> writers address)
+        writers: Map<u64, ContractAddress>,
```
Comment (m-kus) [inline contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:14]: Ig you can use [`Vec`](https://github.com/starkware-libs/cairo/blob/main/corelib/src/starknet/storage/vec.cairo) here

File: contracts/src/contracts/nullifier_registry/nullifier_registry.cairo [hunk 2]
```diff
@@ -1 +1,294 @@
...
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        NullifierAdded: NullifierAdded,
+        WriterAdded: WriterAdded,
+        WriterRemoved: WriterRemoved
+    }
+
+    /// @notice Emitted when the account is locked
```
Comment (m-kus) [inline contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:29]: Docstring seems to be from some different function?

Comment (mubarak23) [reply contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:29]: woo, i will remove it

File: contracts/src/contracts/nullifier_registry/nullifier_registry.cairo [hunk 3]
```diff
@@ -1 +1,294 @@
...
+            assert(self.is_writer(get_caller_address()), 'Caller is not a writer');
+            assert(!self.is_nullified(nullifier), 'Nullifier already exists');
+
+            self.is_nullified.write(nullifier, true);
+            // emit event here
+            self.emit(NullifierAdded { nullifier: nullifier, writer: get_caller_address(), });
+        }
+
+
+        fn add_write_permissions(ref self: ContractState, new_writer: ContractAddress) {
```
Comment (m-kus) [inline contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:68]: This is only allowed for the contract owner, you should use [`Ownable`](https://github.com/OpenZeppelin/cairo-contracts/blob/main/packages/access/src/ownable/ownable.cairo) similarly to how it's done in the original

File: contracts/src/contracts/nullifier_registry/nullifier_registry.cairo [hunk 4]
```diff
@@ -1 +1,294 @@
...
+            assert(!self.is_writer(new_writer), 'The Address is Already a writer');
+            self.is_writers.write(new_writer, true);
+            self.writers.write(self.writers_count.read(), new_writer);
+            self.writers_count.write(self.writers_count.read() + 1);
+            // emit event
+            self.emit(WriterAdded { writer: new_writer });
+        }
+
+
+        fn remove_writer_permissions(ref self: ContractState, remove_writer: ContractAddress) {
```
Comment (m-kus) [inline contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:78]: Same, only owner is allowed

File: contracts/src/contracts/nullifier_registry/nullifier_registry.cairo (hunk 1 above)
Comment (mubarak23) [reply contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:14]: @m-kus can give me an example here, 

i try using this 

` writers: Vec<ContractAddress>,`

and perform appending the address 

`self.all_writers.append(new_writer);`

am getting the following errors

```
Method `append` could not be called on type `core::starknet::storage::storage_base::StorageBase::<core::starknet::storage::Mutable::<core::starknet::storage::vec::Vec::<core::starknet::contract_address::ContractAddress>>>`.
Candidate `ArrayTrait::append` inference failed with: Type mismatch: `core::starknet::storage::storage_base::StorageBase::<core::starknet::storage::Mutable::<core::starknet::storage::vec::Vec::<core::starknet::contract_address::ContractAddress>>>` and `core::array::Array::<?30>`.
Candidate `ByteArrayTrait::append` inference failed with: Type mismatch: `core::starknet::storage::storage_base::StorageBase::<core::starknet::storage::Mutable::<core::starknet::storage::vec::Vec::<core::starknet::contract_address::ContractAddress>>>` and `core::byte_array::ByteArray`.
```

Comment (m-kus) [reply contracts/src/contracts/nullifier_registry/nullifier_registry.cairo:14]: Probably need to import `MutableVecTrait`

**Comment by mubarak23** [comment]
@0xChqrles you can look at it from the recent commit

**Review by 0xChqrles — APPROVED** [review]
lgtm

**Pull request merged**